### ISSUE TITLE
porting joyosc on macos

### ISSUE THREAD

**Issue by fdch** [danomatika/joyosc#6]
I tried making this [package](https://drive.google.com/open?id=18BCzWqWmhMuNXTvHdcPvXDnfrVukvz8r) for joyosc to use on another mac, and I got the following error:
```
dyld: Library not loaded: /usr/local/opt/sdl2/lib/libSDL2-2.0.0.dylib
  Referenced from: /Users/camarahalac.1/Downloads/package/./joyosc
  Reason: image not found
Abort trap: 6
```
Which makes me think that the executable is trying to load its dependencies from their original location obtained from pkg-config, and not from the current directory. I suspect the same thing is happening on Windows (https://github.com/danomatika/joyosc/issues/5), since it's the same type of build. 

In any case, this makes me think that the package won't run easily on other people's computers...

**Comment by danomatika** [comment]
>In any case, this makes me think that the package won't run easily on other people's computers...

Not true. It needs to be built to either 1. look for precompiled dynamic libraries in a local, alternate location or 2. statically link all library dependencies into the executables directly.

**Comment by danomatika** [comment]
Another option is to wrap everything up into .app bundle. The executable would need to be told to look locally with `install_name_tool`.

**Comment by danomatika** [comment]
Actually, it doesn't need to be in a bundle. A script could copy the binaries into a folder with the dynamic libs and then run `install_name_tool` to make sure the binaries look locally first. This would be true for liblo and tinyxml as well as the helper libs in the `lib` folder.

Another option is to statically link everything, but then you end up with `lsjs` and `joyosc` being relatively fat with a full copy of all libs in each.

**Comment by danomatika** [comment]
@agraef has a solution for this, maybe we can bundle a script etc to produce packages on macOS and Windows

**Comment by agraef** [comment]
Yeah, on Mac I basically throw all the binaries along with the required dylibs from homebrew and the documentation files into a directory, add an application bundle which is basically just a shell script with an icon attached to it, and finally package the directory as a dmg file. If you're interested, you can find the latest build for my students here: https://seafile.rlp.net/f/2fb67e3abb1341cf899c/. This isn't quite the latest version yet, but it's pretty recent, and I can hopefully update the package to 1.0 over the weekend.

I haven't fully automatized this process yet, I just update the included binaries and documentation as needed and then use a little script to patch up the library dependencies in the executables, self-sign the package, and create the dmg. FWIW, here's the the script that I use, so that you can build your own packages if you want. It assumes a staging directory with everything that's supposed to end up in the dmg under joyosc/joyosc.

~~~shell
#! /bin/bash

# location of the app bundle
app=joyosc

# fix up the lib dependenciess

for lib in /opt/homebrew/opt/sdl2/lib/libSDL2-2.0.0.dylib; do
    install_name_tool -change $lib @executable_path/$(basename $lib) $app/$app/lsjs
done

for lib in /opt/homebrew/opt/sdl2/lib/libSDL2-2.0.0.dylib /opt/homebrew/opt/liblo/lib/liblo.7.dylib /opt/homebrew/opt/tinyxml2/lib/libtinyxml2.10.dylib; do
    install_name_tool -change $lib @executable_path/$(basename $lib) $app/$app/joyosc
done

# self-sign
codesign --deep --sign "-" $app/$app

# create dmg
rm -f joyosc.dmg
hdiutil create -volname joyosc -srcfolder $app joyosc.dmg
~~~

This is for the Apple Silicon build. Slight adjustments in the homebrew paths would be needed for an Intel build.

**Comment by agraef** [comment]
The Windows package is very similar, but simpler, since I don't have to patch up the dll paths, I just add an application link + icon and zip up the staging directory, that's it. You can find that package here: https://seafile.rlp.net/f/d9513d9981b74aa09e9d/

**Comment by danomatika** [comment]
I'd suggest providing a template config.xml which is used by the wrappers. Then beginners can know to edit this file for customization without having to modify command line options. All options are supported in the XML as well.

**Comment by agraef** [comment]
That's already in the packages that I provide, it includes all the documentation as well, see doc/joyosc in the package.

**Comment by danomatika** [comment]
I'd take a PR for the scripts, if you want. Can put them in a new `scripts` dir with a readme with basic usage info. This is more advanced usage, so I don't think the info needs to be too long or detailed. Another option could be a dedicated `pkg` or `package` dir since the purpose is building and packaging.

**Comment by agraef** [comment]
> I'd take a PR for the scripts

Will do eventually, but I'd rather work on it some more to make the process fully automatic, including integration with the build system and creation of the staging area. We might then use the script in the GH CI to automatically build the Mac and Windows packages and upload those artifacts so that users can just grab the resulting packages from the Action tab. There's also a way to automatically create actual releases from the artifacts whenever a release is tagged. I've done this kind of stuff before, I just need to sit down and work out the details.

**Comment by danomatika** [comment]
Ok, Id say don't make it too complicated. There could be two scripts which can then be added as makefile targets. If each script takes the version as an argument, the package naming can be propagated from the configure version. We do this with Pure Data and `make app`.

**Comment by agraef** [comment]
Well, the script I posted above is all I have right now. Feel free to use that as you see fit.

Anything beyond that means that the script should at least populate the staging area from which the dmg or zip is constructed, which isn't all that complicated. I'm currently working on that.

I'd also like to improve the way that the macOS app bundle is constructed. Currently I use the quick-and-dirty way of just wrapping a shell script which will invoke the joyosc executable, but that makes it hard to include an icon with the bundle. There's an open source application which should help with that, http://www.sveinbjorn.org/platypus, I'd like to give that a try.

**Comment by agraef** [comment]
> I'd also like to improve the way that the macOS app bundle is constructed.

Not necessary to invoke platypus, fileicon makes that very easy (https://github.com/mklement0/fileicon), and it's in homebrew, so I'm using that now. See #29.

**Comment by danomatika** [comment]
@fdch joyosc 1.0.0 is now out and include the make-package.sh script. Please try it out. I will close this for now.
